pilot/pkg/model/push_context.go

```diff
@@ -304,10 +310,21 @@ func (ps *PushContext) UpdateMetrics() {
 	}
 }
 
+// Services returns the list of services that are visible to a Proxy in a given config namespace
+func (ps *PushContext) Services(proxy *Proxy) []*Service {
```

Comment:
My main concern is that will reduce in another regression in performance - this method is called extremely often, and filtering on each call will result in a lot of allocations once filtering is implemented.

That's why I would like the visible services/namespaces to be cached in proxy. I don't mind if  the method is on push context or proxy, but we need to minimize allocs and cache more.

Reply:
There is nothing here that mandates where data has to be stored. WE can store it in the proxy for all we care, or in the push context in some map of namespace, array of services (this is what I was planning to do, instead of having a copy on every proxy struct).

Also, you can only cache the namespace local services in a proxy struct. When someone uses the networkScope to import 6-8 namespaces, we can cache the services list for a namespace (based on network scope) ONCE in push context and just return it here. Instead of storing this info in every proxy.. That was my thinking